File: tools/file-traverse/src/types/file.ts
export type TraverseReq = {
  inputDir: string
  outputDir?: string | null
  organizeByFormat: boolean
  minSizeMb?: number | null
  maxSizeMb?: number | null
  includePatterns: string[]
  excludePatterns: string[]
  /** 仅媒体模式 */
  mediaOnlyMode?: boolean
}

export type FormatStat = { ext: string; scanned: number; count: number; filtered: number }

/** 视频元数据 */
export type VideoMeta = {
  width: number
  height: number
  bitrate: string
  bitrateRaw: number
  aspectRatio: string
  frameRate: string
  frameRateMode: string
  codec: string
  hdrFormat?: string | null
  durationMs: number
  durationStr: string
}

/** 音频元数据 */
export type AudioMeta = {
  codec: string
  sampleRate: string
  bitrate: string
  durationMs: number
  durationStr: string
}

/** 图片元数据 */
export type ImageMeta = {
  width: number
  height: number
  format: string
}

/** 媒体文件信息 */
export type MediaFileInfo = {
  path: string
  name: string
  size: number
  mediaType: 'video' | 'audio' | 'image'
  newName: string
  video?: VideoMeta | null
  audio?: AudioMeta | null
  image?: ImageMeta | null
}

export type TraverseResp = {
  totalFiles: number
  copied: number
  skipped: number
  failed: number
  outputDir: string
  elapsedMs: number
  throughputBytes: number
  formatStats: FormatStat[]
  problems: { path: string; error: string }[]
  /** 媒体模式返回的媒体文件详情 */
  mediaFiles?: MediaFileInfo[]
}

export type PreviewStats = {
  files: { name: string; path: string; size: number }[]
  totalCount: number
  totalSize: number
}

/** 媒体预览请求 */
export type MediaPreviewReq = {
  inputDir: string
}

/** 媒体预览文件（带元数据） */
export type MediaPreviewFile = {
  path: string
  name: string
  ext: string
  size: number
  mediaType: 'video' | 'audio' | 'image'
  video?: VideoMeta | null
  audio?: AudioMeta | null
  image?: ImageMeta | null
}

/** 媒体预览结果 */
export type MediaPreviewResp = {
  videos: MediaPreviewFile[]
  audios: MediaPreviewFile[]
  images: MediaPreviewFile[]
}

File: src-tauri/src/models/file.rs
use serde::{Deserialize, Serialize};
use std::path::PathBuf;

use super::media::MediaFileInfo;

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct TraverseRequest {
  pub input_dir: PathBuf,
  #[serde(default)]
  pub output_dir: Option<PathBuf>,
  #[serde(default)]
  pub organize_by_format: bool,
  #[serde(default)]
  pub min_size_mb: Option<u64>,
  #[serde(default)]
  pub max_size_mb: Option<u64>,
  #[serde(default)]
  pub include_patterns: Vec<String>,
  #[serde(default)]
  pub exclude_patterns: Vec<String>,
  #[serde(default)]
  pub media_only_mode: bool,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct FormatStat {
  pub ext: String,
  pub scanned: usize,
  pub count: usize,
  pub filtered: usize,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct TraverseProblem {
  pub path: PathBuf,
  pub error: String,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct TraverseResult {
  pub total_files: usize,
  pub copied: usize,
  pub skipped: usize,
  pub failed: usize,
  pub output_dir: PathBuf,
  pub elapsed_ms: u128,
  pub throughput_bytes: u64,
  pub format_stats: Vec<FormatStat>,
  pub problems: Vec<TraverseProblem>,
  pub log_path: PathBuf,
  #[serde(default, skip_serializing_if = "Vec::is_empty")]
  pub media_files: Vec<MediaFileInfo>,
}
